synapse/lib/lmdbslab.py

```diff
@@ -1541,109 +1512,124 @@ def bump(self):
             self.curs.close()
             self.bumped = True
 
+    def item(self):
+        return self.curs.item()
+
+    def iterfunc(self, curs):
+        return curs.iternext()
+
+    def resume(self, item):
+
+        if not self.dupsort:
+            return self.curs.set_range(item[0])
+
+        if self.curs.set_range_dup(*item):
+            return True
+
+        if not self.curs.set_range(item[0]):
+            return False
+
+        # if the key is the same, we're at a previous
+        # entry and need to skip dups to the next key
+        if self.curs.key() == item[0]:
+            return self.curs.next_nodup()
+
+        return True
+
+class ScanKeys(Scan):
+
+    def item(self):
+        return self.curs.key()
+
+    def iterfunc(self, curs):
+        return self.curs.iternext(keys=True, values=False)
+
+    def resume(self, item):
+        return self.curs.set_range(item)
+
 class ScanBack(Scan):
     '''
     A state-object used by Slab.  Not to be instantiated directly.
 
     Scans backwards.
     '''
-    def __init__(self, slab, db, singlekey=False):
-        Scan.__init__(self, slab, db, singlekey)
-        if self.singlekey:
-            self.iterfunc = functools.partial(lmdb.Cursor.iterprev_dup, keys=True)
-        else:
-            self.iterfunc = lmdb.Cursor.iterprev
+    def iterfunc(self, curs):
+        return curs.iterprev()
 
     def first(self):
 
         if not self.curs.last():
             return False
 
-        self.genr = self.curs.iterprev()
+        if self.dupsort:
+            self.curs.last_dup()
 
+        self.genr = self.iterfunc(self.curs)
         self.atitem = next(self.genr)
         return True
 
     def set_key(self, lkey):
 
         if not self.curs.set_key(lkey):
-            return False
+            if not self.curs.last():
```

Comment:
What is this for?

Reply:
To move the cursor to the last available entry in the event that set_key() fails.  Since we're iterating backwards, we may specify a key that is past the end of the current db and still want to iterate back

Reply:
Wouldn't you want to be using set_range() to get that behavior?

Reply:
hm.  yea, i think i agree.  Since in the forward case you can iterate past the original key, but you should return False if you fail to land on it directly when set.

Reply:
:+1: